Added Fish completions

faogustavo opened the pull request:
Implementation of #189 

I still need to see how to do the `CompletionCandidates.Custom`. I think the best way to do it is to:
1. Change the `ShellType` from the `CompletionCandidates.Custom` to use the new Shell enum;
2. Use the `ShellType` in the CompletionGenerator;

IMHO we should use the new class to avoid coupling the completion generator with the candidates. But this will be a breaking change.

Btw, here is a gif of the completion working using the code from `samples/plugins`:

| Gif |
| --- |
| ![Kapture 2020-11-11 at 07 58 36](https://user-images.githubusercontent.com/7588480/98803582-c2f82700-23f3-11eb-8d7b-4edea5cd032c.gif) |

ajalt reviewed (changes requested):
Looks great! I just have a few small comments.

There are a few capabilities supported for bash and zsh that would be good to implement for fish if possible:

- Nested subcommands (e.g. `git remote add`).
- Arguments
- `allowInterspersedArgs`
- `CliktCommand.aliases()`

These are not blockers for merging this PR, but would be good to add if any of them are possible with fish.

For `Custom`, let's add a new `FISH` entry to `ShellType`, and switch on that inside `Custom.fromStdout`.

Can you add some tests? You could copy the existing `CompletionTest` and adapt it to fish.

faogustavo commented on clikt/src/commonMain/kotlin/com/github/ajalt/clikt/completion/CompletionGenerator.kt:
```diff
@@ -268,4 +269,146 @@ object CompletionGenerator {
+        val commandName = command.commandName
+        val parentCommandName = parentCommand?.commandName
+
+        val options = command._options //.filterNot { it.hidden }
```
I was making a test and forgot to revert this 😅

faogustavo replied:
BTW, @ajalt should we complete hidden options? I didn't found a filter for that in the bash/zsh function

ajalt replied:
That's a good point. I think we should not complete hidden options. If you want to make that change now, go ahead. Otherwise, I can do it as a separate PR.

faogustavo replied:
Nice. I'll do it here.

faogustavo commented on clikt/src/commonMain/kotlin/com/github/ajalt/clikt/completion/CompletionGenerator.kt:
```diff
@@ -268,4 +269,146 @@ object CompletionGenerator {
+            return ""
+
+        val commandName = command.commandName
+        val needingCommand = "__fish_use_subcommand"
```
In my first test, these variables were specific functions for the script (like in the git completion from fish). I'll replace it

faogustavo commented:
I've already fixed some points and will do the others during this week

faogustavo commented:
@ajalt I've fixed all your comments and added some tests.
From the other points you have mentioned, I've already implemented the `Nested subcommand` by recursion at line 415.
About the arguments, I'm studying how to do it, but it will increase this PR size, in this case, I prefer to open another with that. 
The other points I still need to study a little more the lib and the fish completion patterns.

ajalt commented on clikt/src/commonMain/kotlin/com/github/ajalt/clikt/completion/CompletionGenerator.kt:
```diff
@@ -2,21 +2,36 @@ package com.github.ajalt.clikt.completion
+        return generateCompletion(command, zsh = true)
+    }
+
+    internal fun generateFishCompletion(command: CliktCommand): String {
```
Let's make this public so that people can generate this completion manually.

ajalt commented on clikt/src/jvmTest/kotlin/com/github/ajalt/clikt/completion/CompletionTest.kt:
```diff
@@ -440,4 +440,65 @@ class CompletionTest {
+            val o by option(completionCandidates = CompletionCandidates.Custom.fromStdout("echo foo bar"))
+            val a by argument(completionCandidates = CompletionCandidates.Custom {
+                """
+                WORDS=${'$'}(echo zzz xxx)
+                COMPREPLY=(${'$'}(compgen -W "${'$'}WORDS" -- "${'$'}{COMP_WORDS[${'$'}COMP_CWORD]}"))
```
This isn't fish code?

faogustavo replied:
I just copied the existing test and was studying how to add arguments, forgot to remove this 😅

ajalt commented on clikt/src/commonMain/kotlin/com/github/ajalt/clikt/completion/CompletionGenerator.kt:
```diff
@@ -371,44 +386,49 @@ object CompletionGenerator {
                     is CompletionCandidates.Custom -> {
-                        // TODO: Implement custom fish completion
+                        val customCompletion = completionCandidate.generator(ShellType.FISH)
+                        append("-a $customCompletion ")
```
This doesn't seem to work with multiple lines of fish script. Is it possible to structure this similarly to how bash custom completions are generated? Each custom completion is the body of a fish function, and that function is called to get the completion? I don't know if fish supports that.

faogustavo replied:
As far as I know, here I will just invoke an existing function or use a combination of commands that will return some data to be used in the completion.
It will be something like: `"(__fish_print_hostnames)"` or `"(cat /etc/passwd | cut -d : -f 1)"`

ajalt replied:
Ok. It would be nice to allow multiple line commands like we do with bash, but if it's too much trouble, we can leave it as-is. We'll just have to document the difference.

faogustavo replied:
@ajalt sry for the late reply, looking to others examples, it is possible to have multiple line commands [like this example](https://github.com/fish-shell/fish-shell/blob/master/share/completions/tshark.fish#L25-L28) and multiple lines for manual completion [like this](https://github.com/fish-shell/fish-shell/blob/master/share/completions/tshark.fish#L37-L46), so it will not be a problem.

ajalt commented on clikt/src/commonMain/kotlin/com/github/ajalt/clikt/completion/CompletionGenerator.kt:
```diff
@@ -270,91 +285,91 @@ object CompletionGenerator {
         options.mapNotNull { option ->
             val names = option.names.shortAndLongNames
-            if (names.first == null && names.second == null)
+            if (names.first.isEmpty() && names.second.isEmpty())
```
You'll need to do something like `val names = (option.names + option.secondaryNames).shortandLongNames`, otherwise you miss names for options like `val flag by option().flag("--no-flag")`

ajalt reviewed (changes requested):
Looking good! It's up to you whether you want to add argument support is a separate PR or include it in this one. Either way is fine with me.

ajalt commented:
Thanks for the contribution!
